framework/encode/vulkan_capture_manager.cpp

```diff
@@ -54,31 +55,36 @@ GFXRECON_BEGIN_NAMESPACE(encode)
 
 VulkanCaptureManager* VulkanCaptureManager::instance_ = nullptr;
 LayerTable            VulkanCaptureManager::layer_table_;
+std::mutex            instance_mtx;
 
 bool VulkanCaptureManager::CreateInstance()
 {
-    bool result = CaptureManager::CreateInstance([]() -> CaptureManager* { return instance_; },
-                                                 []() {
-                                                     assert(instance_ == nullptr);
-                                                     instance_ = new VulkanCaptureManager();
-                                                 });
+    std::lock_guard<std::mutex> instance_lock(instance_mtx);
```

Comment:
The original change which pushed the mutex down into `CaptureManager::CreateInstance` was motivated by having multiple possible capture managers (e.g. one for Vulkan and one for DX12).  Raising the mutex back up into the derived class means the two capture managers would no longer protect the critical section in `CaptureManager::CreateInstance`.  @locke-lunarg may have more insight.  I think this needs a different solution.

Reply:
I agreed with Brad. `std::lock_guard<std::mutex> instance_lock(instance_mtx);` stay in CaptureManager::CreateInstanc might be better.

Reply:
I am realizing that this patch is very confusing. Sorry for that.
The changes I did concerning where the locking takes places for the classes of `CaptureManager`/`VulkanCaptureManager` have nothing to do with fixing the deadlock. I got a little carried away and tried to change more things than necessary.

The deadlock happens because `CreateInstance()` in `framework/generated/generated_vulkan_api_call_encoders.cpp` does a `shared_lock` and later down the call stack, in `CaptureManager::CreateCaptureFile()`, each thread will do a (unique)`lock`. This is a deadlock by itself.
Potentially this could be fixed if `CaptureManager::CreateCaptureFile()` first did an `unlock_shared` to release the shared lock before `lock`ing to get exclusive writing access.
But this cannot work in the case of `CreateInstance()` as there is another mutex in `CaptureManager::CreateInstance()`. So the second thread, before it could release its reader's lock, it would get blocked in `CaptureManager::CreateInstance()`.

So a TL;DR version of the above is that the deadlock can easily be fixed by requesting writer's access earlier (in `framework/generated/generated_vulkan_api_call_encoders.cpp` call `AcquireUniqueStateLock()` instead of `AcquireSharedStateLock()`) and not try to change the lock from reader's into writer's.

Now concerning the instance mutex.

The singleton logic we have in place confused me a bit. We have singleton logic in both classes, base and derived, while only the derived class can be instantiated. So I thought moving the singleton mechanism in the derived class would make sense. Will revert the changes back.